### PR TITLE
feat: add Kerberos authentication for Kafka

### PR DESCRIPTION
**Pull request by adwk67**
Fixes https://github.com/stackabletech/kafka-operator/issues/655
Pending decision: https://github.com/stackabletech/decisions/issues/28 ( :heavy_check_mark:  closed/done)

Openshift/OKD tests :green_circle: :

```
--- PASS: kuttl (520.16s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/configuration_kafka-latest-3.7.1_zookeeper-latest-3.9.2_openshift-true (41.05s)
        --- PASS: kuttl/harness/kerberos_kafka-3.8.0_zookeeper-latest-3.9.2_openshift-true_krb5-1.21.1_kerberos-realm-PROD.MYCORP_kerberos-backend-mit (100.10s)
        --- PASS: kuttl/harness/delete-rolegroup_kafka-3.8.0_zookeeper-latest-3.9.2_openshift-true (71.71s)
        --- PASS: kuttl/harness/smoke_kafka-3.8.0_zookeeper-3.9.2_use-client-tls-true_openshift-true (107.86s)
        --- PASS: kuttl/harness/logging_kafka-3.8.0_zookeeper-latest-3.9.2_openshift-true (117.43s)
        --- PASS: kuttl/harness/cluster-operation_kafka-latest-3.7.1_zookeeper-latest-3.9.2_openshift-true (130.04s)
        --- PASS: kuttl/harness/tls_kafka-3.8.0_zookeeper-latest-3.9.2_use-client-tls-true_use-client-auth-tls-true_openshift-true (164.68s)
        --- PASS: kuttl/harness/upgrade_zookeeper-3.9.2_upgrade_old-3.7.1_upgrade_new-3.8.0_use-client-tls-true_use-client-auth-tls-true_openshift-true (151.38s)
PASS
```

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

# Author
- [x] Changes are OpenShift compatible
- [x] CRD changes approved. Awaiting https://github.com/stackabletech/operator-rs/pull/880.
- [x] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [x] Helm chart can be installed and deployed operator works
- [x] Integration tests passed (for non trivial changes)
- [x] Changes need to be "offline" compatible

# Reviewer
- [x] Code contains useful comments
- [x] Code contains useful logging statements
- [x] (Integration-)Test cases added
- [x] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [x] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)

# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated

### CODE REVIEW

**Comment by adwk67** [comment]
@sbernauer I tried to use one principal for all brokers, as we discussed, but could not quite get it to work.
This variation works with the brokers (they all come up):
`principal=\"kafka/test-kafka-broker-default.kuttl-test-glorious-airedale.svc.cluster.local@$KERBEROS_REALM\"`
but I was not able to do the same for the client job. I set 
`BROKER=test-kafka-broker-default.$NAMESPACE.svc.cluster.local:9093` 
which allows me to query the topics but it fails with authentication errors when in producer/consumer mode.

File: rust/crd/src/authentication.rs
Comment (sbernauer) [inline rust/crd/src/authentication.rs:48]: Marker to self: Is this really activating TLS? Where do the certs come from? Is this maybe a copy/paste leftover? :)

File: rust/crd/src/authentication.rs
Comment (sbernauer) [inline rust/crd/src/authentication.rs:132]: You can combine the blocks as they all contain the same content
```suggestion
                AuthenticationClassProvider::Static(_)
                | AuthenticationClassProvider::Ldap(_)
                | AuthenticationClassProvider::Oidc(_) => {
```

File: rust/crd/src/authentication.rs
Comment (sbernauer) [inline rust/crd/src/authentication.rs:133]: Can you please have a look at the TODO from line 21?
> // TODO: Adapt message if multiple authentication classes are supported

File: rust/crd/src/lib.rs
Comment (sbernauer) [inline rust/crd/src/lib.rs:331]:
```suggestion
    /// We only have one role and will use "kafka" everywhere (which e.g. differs from the current hdfs implementation,
    /// but is similar to HBase).
```

File: rust/crd/src/listener.rs
Comment (sbernauer) [inline rust/crd/src/listener.rs:27]: nit: Please spread the enum variants apart using newlines

File: rust/crd/src/listener.rs
Comment (sbernauer) [inline rust/crd/src/listener.rs:117]: optional: I think it would be great if we can assert this, e.g. using
```
        ensure!(
            kafka_security.tls_client_authentication_class().is_none(),
            FooErrorSnafu
        );
```

File: rust/crd/src/security.rs
Comment (sbernauer) [inline rust/crd/src/security.rs:46]: It would be even better top use `SUPPORTED_AUTHENTICATION_CLASS_PROVIDERS` here
```suggestion
    #[snafu(display("only one authentication method is possible at a time, either TLS or Kerberos"))]
```

File: rust/crd/src/security.rs
Comment (sbernauer) [inline rust/crd/src/security.rs:615]: I would prefer if we can get the principal to the following for consistency reasons with other product
`sasl.kerberos.principal=kafka/{kafka_name}@$KERBEROS_REALM`.
I'm happy to try that out as well, maybe a second pair of eyes can find a way to get that working.
But let's merge listener-op in first, it might change things

File: rust/operator-binary/src/kafka_controller.rs
Comment (sbernauer) [inline rust/operator-binary/src/kafka_controller.rs:455]: One nice thing about tracing is that you can capture variables in a structured way, you don't need to squash everything into a string. I  can recommend https://docs.rs/tracing/latest/tracing/#recording-fields
```suggestion
    tracing::debug!(
        kerberos_enabled = kafka_security.has_kerberos_enabled(),
        kerberos_secret_class = ?kafka_security.kerberos_secret_class(),
        tls_enabled = kafka_security.tls_enabled(),
        tls_client_authentication_class = ?kafka_security.tls_client_authentication_class(),
        "The following security settings are used"
    );
```

File: rust/operator-binary/src/kafka_controller.rs
Comment (sbernauer) [inline rust/operator-binary/src/kafka_controller.rs:727]: Same here
```suggestion
    tracing::debug!(?server_cfg, "Applied server config");
    tracing::debug!(?jvm_sec_props, "Applied JVM config");
```

File: rust/operator-binary/src/kerberos.rs
Comment (sbernauer) [inline rust/operator-binary/src/kerberos.rs:46]: Either we get the `kafka/<kafka-name>@REALM` principal working or we should be able to remove the `service` scope

**Review by sbernauer — DISMISSED** [review]
Only did a quick review of the code. I can go into more detail when we merged the changes from https://github.com/stackabletech/kafka-operator/pull/443 and I find the proper time

File: rust/operator-binary/src/kafka_controller.rs
Comment (adwk67) [reply rust/operator-binary/src/kafka_controller.rs]: Thanks! Will take a look at that

File: rust/crd/src/listener.rs
Comment (adwk67) [reply rust/crd/src/listener.rs:28]: a88146c

File: rust/crd/src/authentication.rs
Comment (adwk67) [reply rust/crd/src/authentication.rs:123]: c32b828

File: rust/crd/src/security.rs
Comment (adwk67) [reply rust/crd/src/security.rs]: c32b828

File: rust/crd/src/listener.rs
Comment (sbernauer) [reply rust/crd/src/listener.rs:121]: You are right! We could add a different ensure!() instead, but it's not important. So feel free to ignore :)

File: rust/crd/src/authentication.rs
Comment (adwk67) [reply rust/crd/src/authentication.rs:49]: Fixed when re-working tests e.g. 2e20d5a

File: rust/crd/src/security.rs
Comment (sbernauer) [inline rust/crd/src/security.rs:227]: Did we need to change this because something was broken? I would have expected that it's legit to say "when you enable mutual tls, you also need to enable tls"
If I'm right, we would either need to reword the error message or have to different errors and ensure!()

Comment (adwk67) [reply rust/crd/src/security.rs:227]: One of the upgrade test fails because it sets internal to tls but serverSecret is null.

Comment (adwk67) [reply rust/crd/src/security.rs:227]: I'll have another look and see which one it was.

File: rust/operator-binary/src/kerberos.rs
Comment (sbernauer) [inline rust/operator-binary/src/kerberos.rs:74]: As we do the same thing to both ContainerBuilder we can do that in a loop
```suggestion

        for cb in [cb_kafka, cb_kcat_prober] {
            cb.add_volume_mount("kerberos", STACKABLE_KERBEROS_DIR)
                .context(AddVolumeMountSnafu)?;
            cb.add_env_var("KRB5_CONFIG", STACKABLE_KERBEROS_KRB5_PATH);
            cb.add_env_var(
                "KAFKA_OPTS",
                format!("-Djava.security.krb5.conf={STACKABLE_KERBEROS_KRB5_PATH}",),
            );
        }
```

File: rust/crd/src/security.rs
Comment (adwk67) [reply rust/crd/src/security.rs:227]: Yes, this is expected behaviour but we can re-visit this in a separate [issue](https://github.com/stackabletech/kafka-operator/issues/767).

Comment (sbernauer) [reply rust/crd/src/security.rs:227]: Thanks! Happy to resolve the thread

**Review by siegfriedweber — CHANGES_REQUESTED** [review]

File: rust/crd/src/lib.rs
Comment (adwk67) [reply rust/crd/src/lib.rs]: Yes, just serde testing. i'll delete it as it is misleading too (since you can't have more than one authenticationClass).
3aa923a

File: rust/crd/src/lib.rs
Comment (adwk67) [reply rust/crd/src/lib.rs:340]: No, but I don't think we have done that for hive or hbase, either.

File: rust/operator-binary/src/kafka_controller.rs
Comment (adwk67) [reply rust/operator-binary/src/kafka_controller.rs]: 5b6b7de

File: rust/crd/src/listener.rs
Comment (adwk67) [reply rust/crd/src/listener.rs]: See 2d4feac. I've combined them, but have specified the secure port rather than the function, as this was place that internal tls was enforced for kerberos, rather than relying on `self.tls_internal_secret_class()` having already been set.

File: rust/crd/src/security.rs
Comment (adwk67) [reply rust/crd/src/security.rs:223]: I will try and combine them.

File: rust/crd/src/security.rs
Comment (adwk67) [reply rust/crd/src/security.rs:610]: Not sure if this should be our final solution, but the test now works when using broker listeners instead of the bootstrap: cb8301d.

File: rust/operator-binary/src/kerberos.rs
Comment (adwk67) [reply rust/operator-binary/src/kerberos.rs]: See cb8301d.

**Comment by adwk67** [comment]
Ran nightly suite again locally following latest changes :green_circle: 
```
--- PASS: kuttl (884.94s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/delete-rolegroup_kafka-3.8.0_zookeeper-latest-3.9.2_openshift-false (73.35s)
        --- PASS: kuttl/harness/tls_kafka-3.8.0_zookeeper-latest-3.9.2_use-client-tls-true_use-client-auth-tls-false_openshift-false (177.87s)
        --- PASS: kuttl/harness/upgrade_zookeeper-3.9.2_upgrade_old-3.7.1_upgrade_new-3.8.0_use-client-tls-true_use-client-auth-tls-false_openshift-false (178.71s)
        --- PASS: kuttl/harness/kerberos_kafka-3.8.0_zookeeper-latest-3.9.2_openshift-false_krb5-1.21.1_kerberos-realm-PROD.MYCORP_kerberos-backend-mit (95.47s)
        --- PASS: kuttl/harness/kerberos_kafka-3.8.0_zookeeper-latest-3.9.2_openshift-false_krb5-1.21.1_kerberos-realm-CLUSTER.LOCAL_kerberos-backend-mit (105.52s)
        --- PASS: kuttl/harness/upgrade_zookeeper-3.9.2_upgrade_old-3.7.1_upgrade_new-3.8.0_use-client-tls-true_use-client-auth-tls-true_openshift-false (125.30s)
        --- PASS: kuttl/harness/smoke_kafka-3.8.0_zookeeper-3.9.2_use-client-tls-true_openshift-false (141.44s)
        --- PASS: kuttl/harness/tls_kafka-3.8.0_zookeeper-latest-3.9.2_use-client-tls-false_use-client-auth-tls-true_openshift-false (148.59s)
        --- PASS: kuttl/harness/tls_kafka-3.8.0_zookeeper-latest-3.9.2_use-client-tls-false_use-client-auth-tls-false_openshift-false (65.18s)
        --- PASS: kuttl/harness/logging_kafka-3.8.0_zookeeper-latest-3.9.2_openshift-false (91.36s)
        --- PASS: kuttl/harness/cluster-operation_kafka-latest-3.7.1_zookeeper-latest-3.9.2_openshift-false (88.81s)
        --- PASS: kuttl/harness/configuration_kafka-latest-3.7.1_zookeeper-latest-3.9.2_openshift-false (28.44s)
        --- PASS: kuttl/harness/smoke_kafka-3.8.0_zookeeper-3.9.2_use-client-tls-false_openshift-false (58.76s)
        --- PASS: kuttl/harness/upgrade_zookeeper-3.9.2_upgrade_old-3.7.1_upgrade_new-3.8.0_use-client-tls-false_use-client-auth-tls-false_openshift-false (120.38s)
        --- PASS: kuttl/harness/tls_kafka-3.8.0_zookeeper-latest-3.9.2_use-client-tls-true_use-client-auth-tls-true_openshift-false (160.60s)
        --- PASS: kuttl/harness/upgrade_zookeeper-3.9.2_upgrade_old-3.7.1_upgrade_new-3.8.0_use-client-tls-false_use-client-auth-tls-true_openshift-false (82.92s)
```

File: rust/crd/src/security.rs
Comment (adwk67) [reply rust/crd/src/security.rs:663]: See c852a1b for changes with new listener for bootstrapper.

File: rust/crd/src/security.rs
Comment (adwk67) [reply rust/crd/src/security.rs]: See 57b7ffc.

File: rust/crd/src/security.rs
Comment (adwk67) [reply rust/crd/src/security.rs:641]: I'd like to do this as part of a separate PR, as this one has already expanded in scope. Would you be in agreement?

**Comment by adwk67** [comment]
Re-[tested](https://testing.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/kafka-operator-it-custom/11/) :green_circle:

File: rust/crd/src/security.rs
Comment (siegfriedweber) [reply rust/crd/src/security.rs:646]: I'm okay with it.

Comment (adwk67) [reply rust/crd/src/security.rs:646]: Covered by https://github.com/stackabletech/kafka-operator/issues/779

File: tests/templates/kuttl/kerberos/30-access-kafka.txt.j2
Comment (siegfriedweber) [inline tests/templates/kuttl/kerberos/30-access-kafka.txt.j2:79]: The discovery ConfigMap `test-kafka` contains `test-kafka-broker-default-bootstrap.kuttl.svc.cluster.local:9093`. It should contain the correct service and port (in this case 9095) and it should be used in the tests.

File: rust/crd/src/security.rs
Comment (siegfriedweber) [inline rust/crd/src/security.rs:76]: There are now a lot of ports and it is not obvious for the user which one to use. The documentation should clarify this.

**Review by siegfriedweber — CHANGES_REQUESTED** [review]

File: rust/crd/src/security.rs
Comment (adwk67) [reply rust/crd/src/security.rs:80]: See 109ad3c

File: tests/templates/kuttl/kerberos/30-access-kafka.txt.j2
Comment (adwk67) [reply tests/templates/kuttl/kerberos/30-access-kafka.txt.j2:79]: Should be fixed in 109ad3c (tested with kerberos and non-kerberos cases).

Comment (adwk67) [reply tests/templates/kuttl/kerberos/30-access-kafka.txt.j2]: Test now uses discovery: [6ced9a5](https://github.com/stackabletech/kafka-operator/pull/762/commits/6ced9a5cacdb84ede039144cb66719bfd7574f7a)

File: rust/crd/src/security.rs
Comment (adwk67) [reply rust/crd/src/security.rs:80]: I thought you meant in the code, but did you actually mean the user docs?

**Review by siegfriedweber — APPROVED** [review]
LGTM

File: rust/crd/src/security.rs
Comment (adwk67) [reply rust/crd/src/security.rs:80]: See c9336ed

File: docs/modules/kafka/pages/usage-guide/security.adoc
Comment (siegfriedweber) [inline docs/modules/kafka/pages/usage-guide/security.adoc:163]:
```suggestion
In order for this to work, it is necessary for kerberized clusters to define an extra Kafka listener for the bootstrap with a corresponding service (and port).
```

File: docs/modules/kafka/pages/usage-guide/security.adoc
Comment (siegfriedweber) [inline docs/modules/kafka/pages/usage-guide/security.adoc:164]: Port 9094 is currently not used. Should we mention this? But I am also okay with the text as it is written now.

**Review by siegfriedweber — CHANGES_REQUESTED** [review]
Good explanation!

File: docs/modules/kafka/pages/usage-guide/security.adoc
Comment (adwk67) [reply docs/modules/kafka/pages/usage-guide/security.adoc]: I've clarified it a bit: 74f2cb8

**Review by siegfriedweber — APPROVED** [review]

**Comment by adwk67** [comment]
Re-testing before merge:

- AKS :green_circle: https://testing.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/kafka-operator-it-custom/12/
- OKD :orange_circle: https://testing.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/kafka-operator-it-custom/13/

**Pull request merged**
